Arrangement_on_surface_2/include/CGAL/draw_arrangement_2.h

```diff
@@ -0,0 +1,588 @@
+    {
+      const auto* traits = this->m_arr.geometry_traits();
+      auto approx = traits->approximate_2_object();
+      draw_approximate_curve(curve, approx);
+      return;
+    }
+    draw_exact_curve(curve);
+#else
+    const auto* traits = this->m_arr.geometry_traits();
+    draw_curve_impl1(curve, *traits, 0);
+#endif
+  }
+
+  /*! Compile time dispatching
+   */
+  template <typename T>
+  auto draw_point_impl2(const Point& p, T const& approx, long) ->
+    decltype(void())
+  { add_point(p); }
+
+  template <typename T>
+  auto draw_point_impl2(const Point& p, T const& approx, int) ->
+    decltype(approx.operator()(p), void())
+  { add_point(approx(p)); }
+
+  template <typename T>
+  auto draw_point_impl1(const Point& p, T const& traits, long) ->
+    decltype(void())
+  { add_point(p); }
+
+  template <typename T>
+  auto draw_point_impl1(const Point& p, T const& traits, int) ->
+    decltype(traits.approximate_2_object(), void()) {
+    using Approximate = typename Gt::Approximate_2;
+    draw_point_impl2<Approximate>(p, traits.approximate_2_object(), 0);
+  }
+
+  /*! Draw a point.
+   */
+  void draw_point(const Point& p) {
+    const auto* traits = m_arr.geometry_traits();
+    draw_point_impl1(p, *traits, 0);
+  }
+
+  /*! Add a Connected Component of the Boundary.
+   */
+  void add_ccb(Ccb_halfedge_const_circulator circ) {
+    auto curr = circ;
+    do {
+      auto new_face = curr->twin()->face();
+      if (m_visited.find(new_face) != m_visited.end()) continue;
+      m_visited[new_face] = true;
+      add_face(new_face);
+    } while (++curr != circ);
+  }
+
+  /*! Add a face.
+   */
+  void add_face(Face_const_handle face) {
+    for (auto it = face->inner_ccbs_begin(); it != face->inner_ccbs_end(); ++it)
+      add_ccb(*it);
+
+    for (auto it = face->outer_ccbs_begin(); it != face->outer_ccbs_end(); ++it)
+    {
+      add_ccb(*it);
+      draw_region(*it);
+    }
+  }
+
+  //!
+  virtual void keyPressEvent(QKeyEvent* e) {
+    // Test key pressed:
+    //    const ::Qt::KeyboardModifiers modifiers = e->modifiers();
+    //    if ((e->key()==Qt::Key_PageUp) && (modifiers==Qt::NoButton)) { ... }
+
+    // Call: * add_elements() if the model changed, followed by
+    //       * redraw() if some viewing parameters changed that implies some
+    //                  modifications of the buffers
+    //                  (eg. type of normal, color/mono)
+    //       * update() just to update the drawing
+
+    // Call the base method to process others/classicals key
+    Base::keyPressEvent(e);
+  }
+
+protected:
+  //! The window width in pixels.
+  int m_width = CGAL_BASIC_VIEWER_INIT_SIZE_X;
+
+  //! The window height in pixels.
+  int m_height = CGAL_BASIC_VIEWER_INIT_SIZE_Y;
+
+  //! The ratio between pixel and opengl units (in world coordinate system).
+  double m_pixel_ratio = 1;
+
+  //! The arrangement to draw.
+  const Arr& m_arr;
+
+  //! The color generator.
+  Color_generator& m_color_generator;
```

Comment:
Did you really mean to put a reference here?  In the constructor it is init with a temporary so there is something fishy.

Reply:
No, I did not. Thanks for spotting it.
I fixed it and added (to the example program draw_arr.cpp) code that
defines a color generator.